Content.Server/Medical/DefibrillatorSystem.cs

```diff
@@ -0,0 +1,258 @@
+            {
+                BlockDuplicate = true,
+                BreakOnUserMove = true,
+                BreakOnTargetMove = true,
+                BreakOnHandChange = true,
+                NeedHand = true
+            });
+    }
+
+    public void Zap(EntityUid uid, EntityUid target, EntityUid user, DefibrillatorComponent? component = null, MobStateComponent? mob = null, MobThresholdsComponent? thresholds = null)
+    {
+        if (!Resolve(uid, ref component) || !Resolve(target, ref mob, ref thresholds, false))
+            return;
+
+        // clowns zap themselves
+        if (HasComp<ClumsyComponent>(user) && user != target)
+        {
+            Zap(uid, user, user, component, mob, thresholds);
+            return;
+        }
+
+        if (!_powerCell.TryUseActivatableCharge(uid, user: user))
+            return;
+
+        _mobThreshold.SetAllowRevives(target, true, thresholds);
+        _audio.PlayPvs(component.ZapSound, uid);
+        _electrocution.TryDoElectrocution(target, null, component.ZapDamage, component.WritheDuration, true, ignoreInsulation: true);
+
+        if (_mobState.IsIncapacitated(target, mob))
+            _damageable.TryChangeDamage(target, component.ZapHeal, true, origin: uid);
+
+        component.NextShockTime = _timing.CurTime + component.ShockDelay;
+        component.CooldownEnded = false;
+        _appearance.SetData(uid, DefibrillatorVisuals.Ready, false);
+        _mobState.ChangeMobState(target, MobState.Critical, mob, uid);
+        _mobThreshold.SetAllowRevives(target, false, thresholds);
+
+        IPlayerSession? session = null;
+        if (TryComp<MindComponent>(target, out var mindComp) &&
+            mindComp.Mind?.UserId != null &&
+            _playerManager.TryGetSessionById(mindComp.Mind.UserId.Value, out session))
+        {
+            // notify them they're being revived.
+            if (mindComp.Mind.CurrentEntity != target)
+            {
+                _chatManager.TrySendInGameICMessage(uid, Loc.GetString("defibrillator-ghosted"),
+                    InGameICChatType.Speak, true, true);
+                _euiManager.OpenEui(new ReturnToBodyEui(mindComp.Mind), session);
+            }
+        }
+        else
+        {
+            _chatManager.TrySendInGameICMessage(uid, Loc.GetString("defibrillator-no-mind"),
+                InGameICChatType.Speak, true, true);
+        }
+
+        var sound = _mobState.IsAlive(target, mob) && session != null
+            ? component.SuccessSound
+            : component.FailureSound;
+        _audio.PlayPvs(sound, uid);
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<DefibrillatorComponent>();
+        while (query.MoveNext(out var uid, out var defib))
+        {
+            if (defib.CooldownEnded)
+                continue;
+
+            if (_timing.CurTime < defib.NextShockTime)
+                continue;
```

Comment:
Is CooldownEnded needed? Can't you just infer it from currenttime < nextshocktime?